opal/util/info_subscriber.c

```diff
@@ -110,41 +109,32 @@ static void opal_callback_list_item_destruct(opal_callback_list_item_t *obj)
     }
 }
 
-static const char *opal_infosubscribe_inform_subscribers(opal_infosubscriber_t *object,
-                                                         const char *key, const char *new_value,
-                                                         int *found_callback)
+static int opal_infosubscribe_inform_subscribers(opal_infosubscriber_t *object, const char *key,
+                                                 const char *new_value)
 {
     opal_hash_table_t *table = &object->s_subscriber_table;
     opal_list_t *list = NULL;
     opal_callback_list_item_t *item;
-    const char *updated_value = NULL;
+    int update_cnt = 0;
 
-    if (found_callback) {
-        *found_callback = 0;
-    }
     /*
      * Present the new value to each subscriber.  They can decide to accept it, ignore it, or
      * over-ride it with their own value (like ignore, but they specify what value they want it to
      * have).
-     *
-     * Since multiple subscribers could set values, only the last setting is kept as the
-     * returned value.
      */
     if (table) {
         opal_hash_table_get_value_ptr(table, key, strlen(key), (void **) &list);
 
         if (list) {
-            updated_value = new_value;
             OPAL_LIST_FOREACH (item, list, opal_callback_list_item_t) {
-                updated_value = item->callback(object, key, updated_value);
-                if (found_callback) {
-                    *found_callback = 1;
+                if (item->callback(object, key, new_value)) {
```

Comment:
Here we are throwing away the value returned from the callback. I assume this is because we are assuming that the potentially updated key will be saved in `new_value`? But seeing some instances where a subscriber is utilized (`mca_pml_ob1_set_allow_overtake`, `ompi_osc_ucx_set_no_lock_info`), the opposite appears to happen? (this looks to be a result of somewhat confusing variable naming in the original implementation?)

Reply:
@wenduwan any thoughts on this? I'm looking at e.g. `mca_pml_ob1_set_allow_overtake()`, which appears to be an info-subscriber callback (registered in `mca_pml_ob1_add_comm()`):
https://github.com/open-mpi/ompi/blob/bb7ecde1d52d0c1f058d2585e9a1f9794df40684/ompi/mca/pml/ob1/pml_ob1.c#L209-L243
https://github.com/open-mpi/ompi/blob/bb7ecde1d52d0c1f058d2585e9a1f9794df40684/ompi/mca/pml/ob1/pml_ob1.c#L268-L270

The modified info value is returned from the function. From what I understand we need to keep the return value and pass *it* to `opal_info_set()`  in `opal_infosubscribe_change_info()`.

Reply:
@gkatev Many apologies for missing your comment earlier.

I took another look at this change. I don't quite understand how the callback value can be useful in this case(and now I doubt if this change is even correct).

Consider this snippet:
```c
OPAL_LIST_FOREACH (item, list, opal_callback_list_item_t) {
        updated_value = item->callback(object, key, updated_value);
        ...
}
```
This code iterates over a list of subscribers, and invokes the callback with a **mutable** variable `updated_value`, and after the loop `updated_value` will be set to the last callback's return value. I have 2 problems/questions about this approach:
1. Why would it invoke the callback with a changing `update_value` - this means the order of subscribers matters. I do not understand this.
2. Why would it only keep the last `updated_value` and ignore the others? Again this means the order of subscribers must be critical.

@bosilca @devreal Could you please give some background about the original purpose of `opal_infosubscribe_inform_subscribers`?

Reply:
The MPI info values are complicated to handle, because they need to reflect not only what the user intent but also what the library accept. Thus, in some cases the user should be able to investigate which of the info provided will be taken into consideration by the implementation. This mechanism is set in place in OMPI via the subscribers, when the user sets a value we inform the subscribers about it and then we set it with a flag to indicate if the it was accepted or not.

Let's add this into the next week's agenda, as it deserves a more in-depth discussion.

Reply:
Added to meeting agenda

Reply:
> This code iterates over a list of subscribers, and invokes the callback with a **mutable** variable `updated_value`, and after the loop `updated_value` will be set to the last callback's return value. I have 2 problems/questions about this approach:
> 
>     1. Why would it invoke the callback with a changing `update_value` - this means the order of subscribers matters. I do not understand this.
> 
>     2. Why would it only keep the last `updated_value` and ignore the others? Again this means the order of subscribers must be critical.

Hmm I see yes. With the current state of the callbacks it generally seems like we have to keep and use the return value, but your concerns above do indeed raise deeper questions about how all this is supposed to work (and I personally don't really know 🙂),

Reply:
Because it was never meant to support multiple subscribers on the same info key. Basically the idea is that an info object belongs, or could impact, a single subscriber. Which means that is a single subscriber that will modify the value, while everyone else might just look at it. The order might matter, but clearly we never had an issue with the ordering so far.